libs/vkd3d/command.c

```diff
@@ -18044,6 +17976,88 @@ static bool d3d12_command_queue_needs_cpu_waits_locked(struct d3d12_command_queu
+
+    command_queue->wait_fence_count = 0;
+}
+
+static void d3d12_command_queue_push_fence_waits_to_worker(struct d3d12_command_queue *command_queue)
+{
+    struct vkd3d_fence_worker *worker = &command_queue->fence_worker;
+    const struct vkd3d_fence_virtual_wait *fence_wait;
+    struct vkd3d_queue_timeline_trace_cookie cookie;
+    struct vkd3d_fence_wait_info fence_info;
+    HRESULT hr;
+    size_t i;
+
+    for (i = 0; i < command_queue->wait_fence_count; i++)
+    {
+        fence_wait = &command_queue->wait_fences[i];
+        cookie = vkd3d_queue_timeline_trace_register_wait(&worker->device->queue_timeline_trace,
+                &fence_wait->fence->ID3D12Fence_iface, fence_wait->virtual_value);
+
+        memset(&fence_info, 0, sizeof(fence_info));
+        fence_info.fence = &command_queue->wait_fences[i].fence->ID3D12Fence_iface;
+        fence_info.vk_semaphore = fence_wait->vk_semaphore;
+        fence_info.vk_semaphore_value = fence_wait->vk_semaphore_value;
+
+        if (FAILED(hr = vkd3d_enqueue_timeline_semaphore(worker, &fence_info, &cookie)))
+            ERR("Failed to enqueue timeline semaphore, hr %#x.\n", hr);
+    }
+}
+
+static void d3d12_command_queue_add_wait(struct d3d12_command_queue *command_queue,
+        struct d3d12_fence *fence, const struct d3d12_fence_value *fence_value)
+{
+    struct vkd3d_fence_virtual_wait *fence_wait;
+
+    if (!vkd3d_array_reserve((void**)&command_queue->wait_fences, &command_queue->wait_fences_size,
+            command_queue->wait_fence_count + 1, sizeof(*command_queue->wait_fences)))
+    {
+        ERR("Failed to add fence wait to queue.\n");
+        return;
+    }
+
+    fence_wait = &command_queue->wait_fences[command_queue->wait_fence_count++];
+
+    memset(fence_wait, 0, sizeof(*fence_wait));
+    fence_wait->fence = fence;
+    fence_wait->virtual_value = fence_value->virtual_value;
+    fence_wait->vk_semaphore = fence_value->vk_semaphore;
+    fence_wait->vk_semaphore_value = fence_value->vk_semaphore_value;
+
+    d3d12_fence_inc_ref(fence);
+
+    vkd3d_queue_add_wait_locked(command_queue->vkd3d_queue,
+            fence_value->vk_semaphore, fence_value->vk_semaphore_value);
+}
+
+static void d3d12_command_queue_flush_waiters(struct d3d12_command_queue *command_queue)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &command_queue->device->vk_procs;
+    VkQueue vk_queue;
+
+    if (!(vk_queue = vkd3d_queue_acquire(command_queue->vkd3d_queue)))
+    {
+        ERR("Failed to acquire queue %p.\n", command_queue->vkd3d_queue);
+        return;
+    }
+
+    vkd3d_queue_flush_waiters_locked(command_queue->vkd3d_queue, vk_queue, command_queue, vk_procs);
+
+    command_queue->last_submission_timeline_value = command_queue->vkd3d_queue->submission_timeline_count;
+
+    d3d12_command_queue_push_fence_waits_to_worker(command_queue);
+    d3d12_command_queue_reset_fence_waits(command_queue);
+
+    vkd3d_queue_release(command_queue->vkd3d_queue);
```

Comment:
There should be no need to hold the lock here. It can be released after `command_queue->last_submission_timeline_value` assignment I think?

Reply:
Fixed.